### ISSUE TITLE
When an exported variable of a tool script is edited in the editor, setters for all instances of that scene are called

### ISSUE THREAD

**Issue by mitchcurtis** [godotengine/godot#28246]
**Godot version:**
v3.1

**OS/device including version:**
MacBook Pro
macOS 10.14.4

**Issue description:**
I have a scene that represents a brick wall. I have a single brick texture that should repeat, so I figured I could use a tool script to allow the amount of bricks to be set, and then have the setter update the sprite's region_rect and collision shape's extents accordingly.

What actually happens is that, although the sprite repeats as expected, the collision shape extents seem to be modified on _every_ instance of the scene (wall) that was modified:

![Kapture 2019-04-20 at 22 49 39](https://user-images.githubusercontent.com/6002340/56462160-bc8ffd80-63be-11e9-921d-b823124a2589.gif)

**Steps to reproduce:**
Take the same steps shown in the GIF with the attached minimal project below.

**Minimal reproduction project:**
[tool-test.zip](https://github.com/godotengine/godot/files/3100596/tool-test.zip)

**Comment by clayjohn** [comment]
This is because you are sharing the collision shape between instances. You need to select "Make Unique" on one of them. You do that by clicking the little down arrow in the inspector beside the Shape in the CollisionShape node and selecting "Make Unique"

**Comment by mitchcurtis** [comment]
Huh, I had no idea. How does that happen? And why only for the shape and not the sprite?

I created the wall scene by setting it up in the main scene and then saving the static body node branch as a scene, if that helps.

**Comment by mitchcurtis** [comment]
https://github.com/godotengine/godot/issues/3222#issuecomment-168666945 says:

> Sub-resources are always shared automatically between instances, Will
> eventually at some point add a flag so the resources are unique to the
> scene, but the default behavior will not change.
> For duplicated objects, you can use the property editor options to make
> sub-resources unique

But that still doesn't explain why the sprite doesn't behave the same way. As a newcomer to Godot, this is very confusing.

https://github.com/godotengine/godot/issues/10147#issuecomment-350505209 mentions yet another variable:

> If that's the problem then as far as I know you should set 'local to scene' inside material resource.

From reading [this](https://godotengine.org/qa/25521/what-does-the-local-to-scene-option-on-resources-do-mean) answer, it seems that local to scene is identical to making nodes unique?

Despite all of this, https://github.com/godotengine/godot/issues/6922 is still open. How are users expected to know this stuff?

**Comment by mitchcurtis** [comment]
> This is because you are sharing the collision shape between instances. You need to select "Make Unique" on one of them. You do that by clicking the little down arrow in the inspector beside the Shape in the CollisionShape node and selecting "Make Unique"

By the way, I tried this, but because the CollisionShape2D is a child of the root node, I had to do it in the wall scene itself, which had no effect:

<img width="230" alt="Screen Shot 2019-04-21 at 10 35 14 AM" src="https://user-images.githubusercontent.com/6002340/56467598-42dd2b80-6421-11e9-9c3a-85cde31d7044.png">

I just tried selecting one of the wall instances in the main scene and clicking Make Sub-Resources Unique, but that did nothing:

<img width="260" alt="Screen Shot 2019-04-21 at 10 37 10 AM" src="https://user-images.githubusercontent.com/6002340/56467618-8172e600-6421-11e9-87ea-23649c8f2adb.png">

I tried doing it on _both_ of them, but still the shapes are shared.

**Comment by bojidar-bg** [comment]
Since it is still the same scene, no amount of making the shape unique will help.
What you want is checking the "Local to scene" property of the Shape resource, or, alternatively, duplicating the shape manually in your script.

**Comment by mitchcurtis** [comment]
I gave up on this and just have e.g. BrickWall1x.tscn, BrickWall4x.tscn, BrickWall8x.tscn, etc.

It would be great if this could be better documented, but unless someone is going to do that, this may as well be closed.

**Comment by clayjohn** [comment]
Did you try what bojidar-bg suggested? You need to make the shapes unique through GDScript when you instantiate the scene.

**Comment by mitchcurtis** [comment]
> Did you try what bojidar-bg suggested? You need to make the shapes unique through GDScript when you instantiate the scene.

For some reason I thought I had explained why I didn't bother trying it, but I see that I didn't.

My understanding of the approach is that making the shapes local to the scene means that they cannot be shared/that they are duplicated? I'm hesitant to fight against the engine in that sense, especially as I'm still just learning.

**Comment by clayjohn** [comment]
In this case, you need to duplicate the shapes if you want them to have unique properties.

**Comment by mitchcurtis** [comment]
Yeah, I get that, but I'm still unsure if this is a better approach (less resource-intensive) than having BrickWall1x.tscn, BrickWall4x.tscn, BrickWall8x.tscn, etc., where each scene has a shape that is shared among those instances.

For example, if I have:

- Two BrickWall1x.tscn instances
- Five BrickWall8x.tscn instances

It's still just two shapes. Whereas if I do the same except with the exported variable + tool script approach, it seems that each shape in all seven BrickWall.tscn instances will be unique and not shared. Or have I misunderstood it?

**Comment by clayjohn** [comment]
That's right. You'll only have two shapes that way. I'm not sure that it would be much faster though. But, if it works, it works! Games are a big bundle of hacks anyway, so stick with whatever works for your project.

**Comment by bojidar-bg** [comment]
Note that you can probably just put multiple shapes side by side.

**Comment by AdaLollA** [comment]
I have the same issue but with an exportet integer. The Setter of this integer changes all instances of the scene - however it should only update on the scene where the variable was changed.

How would one go about solving this problem? Is there a different way to achieve this goal (except for creating scenes for every single possible variable case as mentioned above)?

**Comment by eranimo** [comment]
Can confirm this is still an issue in Godot v4.0

**Comment by Jaaaas** [comment]
Can confirm this is still an issue in Godot v4.0.
